File: leetcode/medium/424-Longest-Repeating-Character-Replacement.py
#!/usr/bin/python3

class Solution:
  # Intital solution
  def characterReplacement(self, s: str, k: int) -> int:
    freq = {}
    i = 0
    j = 0
    res = 0
    most = 0
    while j < len(s):
      length = j - i + 1 
      freq[s[j]] = freq.get(s[j], 0) + 1
      most = max(freq.values())
      if length - most <= k:
        res = max(length, res)
        j += 1
      else:
        freq[s[i]] -= 1
        freq[s[j]] -= 1
        i += 1
    return res
  
  # Cleaner O(26 * n) solution
  def characterReplacement(self, s: str, k: int) -> int:
    freq = {}
    res = 0
    l = 0
    for r in range(len(s)):
      freq[s[r]] = freq.get(s[r], 0) + 1
      most = max(freq.values()) 
      if r - l + 1 - most <= k:
        res = max(res, r - l + 1)
      else:
        freq[s[l]] -= 1
        l += 1
    return res 

  # O(n) solution using a maxFrequency variable
  def characterReplacement(self, s: str, k: int) -> int:
    freq = {}
    res = 0
    l = 0
    maxf = 0
    for r in range(len(s)):
      freq[s[r]] = freq.get(s[r], 0) + 1
      maxf = max(maxf, freq[s[r]])
      if r - l + 1 - maxf <= k:
        res = max(res, r - l + 1)
      else:
        freq[s[l]] -= 1
        l += 1
    return res 

File: neetcode150/medium/cpp/424-Longest-Repeating-Character-Replacement.cpp
#include <map>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        map<char,int> h;

        int i = 0;
        int j = 0;

        int res = 0;

        int curr = 0;

        while (j < s.size()) {
            auto [it, inserted] = h.try_emplace(s[j], 1);
            if (!inserted) {
                ++it->second;
            }

            curr = max(curr, it->second);

            while ((j-i+1)-curr > k) {
                int& v = h[s[i]];
                --v;
                if (v == 0) { h.erase(s[i]); }
                ++i;
            }

            res = max(res, j-i+1);
            ++j;
        }

        return res;
    }
};